diff -Naur grub-new/util/grub.d/10_linux.in grub-new2/util/grub.d/10_linux.in
--- grub-new/util/grub.d/10_linux.in	2013-02-27 12:53:08.192681839 +0400
+++ grub-new2/util/grub.d/10_linux.in	2013-02-27 12:54:52.196458139 +0400
@@ -31,7 +31,7 @@
 if [ "x${GRUB_DISTRIBUTOR}" = "x" ] ; then
   OS="$(sed 's, release .*$,,g' /etc/system-release) Linux"
 else
-  OS="${GRUB_DISTRIBUTOR} Linux"
+  OS="$(sed 's, release .*$,,g' /etc/system-release) 2012"
   CLASS="--class $(echo ${GRUB_DISTRIBUTOR} | tr 'A-Z' 'a-z' | cut -d' ' -f1) ${CLASS}"
 fi
 
