diff -Nur grub-2.00.old/grub-core/script/yylex.l grub-2.00/grub-core/script/yylex.l
--- grub-2.00.old/grub-core/script/yylex.l	2012-06-09 00:24:15.000000000 +0400
+++ grub-2.00/grub-core/script/yylex.l	2014-07-09 11:35:57.000537711 +0400
@@ -29,6 +29,8 @@
 #pragma GCC diagnostic ignored "-Wmissing-prototypes"
 #pragma GCC diagnostic ignored "-Wmissing-declarations"
 #pragma GCC diagnostic ignored "-Wunsafe-loop-optimizations"
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#pragma GCC diagnostic ignored "-Wunused-function"
 
 #define yyfree    grub_lexer_yyfree
 #define yyalloc   grub_lexer_yyalloc
