diff --git a/util/grub-install.in b/util/grub-install.in
index c108fa5..560e91b 100644
--- a/util/grub-install.in
+++ b/util/grub-install.in
@@ -827,6 +827,15 @@ elif [ x"$grub_modinfo_platform" = xefi ]; then
     if [ x$grub_modinfo_target_cpu = xi386 ]; then
 	cp "${grubdir}/${grub_modinfo_target_cpu}-$grub_modinfo_platform/core.${imgext}" "${efidir}/boot.efi"
     fi
+    # Create minimal grub.cfg to load the main grub.cfg
+    boot_uuid=`$grub_probe --target=fs_uuid /boot 2>/dev/null`
+    boot_hint=`$grub_probe --target=compatibility_hint /boot 2>/dev/null`
+    prefix_dir=`$grub_mkrelpath /boot/grub2`
+    cat <<EOF > "${efidir}/grub.cfg"
+search.fs_uuid $boot_uuid root $boot_hint
+set prefix=(\$root)$prefix_dir
+configfile \$prefix/grub.cfg
+EOF
 
     # Try to make this image bootable using the EFI Boot Manager, if available.
     efibootmgr="`which efibootmgr`"
