diff -Naur grub-1.99//include/grub/term.h grub-new2//include/grub/term.h
--- grub-1.99//include/grub/term.h	2011-04-10 14:11:43.000000000 +0400
+++ grub-new2//include/grub/term.h	2012-06-26 15:40:09.019987653 +0400
@@ -224,9 +224,9 @@
 };
 typedef struct grub_term_output *grub_term_output_t;
 
-#define GRUB_TERM_DEFAULT_NORMAL_COLOR 0x07
-#define GRUB_TERM_DEFAULT_HIGHLIGHT_COLOR 0x70
-#define GRUB_TERM_DEFAULT_STANDARD_COLOR 0x07
+#define GRUB_TERM_DEFAULT_NORMAL_COLOR 0x00
+#define GRUB_TERM_DEFAULT_HIGHLIGHT_COLOR 0x00
+#define GRUB_TERM_DEFAULT_STANDARD_COLOR 0x00
 
 extern struct grub_term_output *EXPORT_VAR(grub_term_outputs_disabled);
 extern struct grub_term_input *EXPORT_VAR(grub_term_inputs_disabled);
