diff --git a/util/grub-install.in b/util/grub-install.in
index 16ff1f6..c108fa5 100644
--- a/util/grub-install.in
+++ b/util/grub-install.in
@@ -863,7 +863,11 @@ elif [ x"$grub_modinfo_platform" = xefi ]; then
 		efi_file_boot="$efi_file"
 	    fi
 	    efidir_part="$(echo "$efidir_drive" | sed 's/^([^,]*,[^0-9]*//; s/[^0-9].*//')"
-	    efibootmgr $efi_quiet -c -d "$efidir_disk" -p "$efidir_part" -w \
+	    disk_id="$(fdisk -l "$efidir_disk" | grep 'Disk identifier:')"
+	    if [ "$disk_id" = "Disk identifier: 0x00000000" ]; then
+		force_write_disk_id="-w"
+	    fi
+	    efibootmgr $efi_quiet -c -d "$efidir_disk" -p "$efidir_part" $force_write_disk_id \
 		-L "$bootloader_id" -l "\\EFI\\$efi_distributor\\$efi_file_boot"
 	fi
     fi
