diff -Naur grub-2.00//util/grub-mkconfig_lib.in grub-new//util/grub-mkconfig_lib.in
--- grub-2.00//util/grub-mkconfig_lib.in	2012-06-27 19:27:53.000000000 +0400
+++ grub-new//util/grub-mkconfig_lib.in	2012-07-23 18:29:36.162108024 +0400
@@ -180,6 +180,9 @@
       *.rpmsave|*.rpmnew) return 1 ;;
       README*|*/README*)  return 1 ;; # documentation
     esac
+    if test -L "$1" ; then
+	return 1
+    fi
   else
     return 1
   fi
