diff -Nur grub-2.00.old/util/grub.d/10_linux.in grub-2.00/util/grub.d/10_linux.in
--- grub-2.00.old/util/grub.d/10_linux.in	2012-04-19 01:24:38.000000000 +0400
+++ grub-2.00/util/grub.d/10_linux.in	2013-08-30 02:16:29.016259085 +0400
@@ -244,6 +244,8 @@
   if [ "x${GRUB_DISABLE_RECOVERY}" != "xtrue" ]; then
     linux_entry "${OS}" "${version}" recovery \
                 "single ${GRUB_CMDLINE_LINUX}"
+    linux_entry "${OS} Safe mode boot" "${version}}" simple \
+    "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT} nomodeset xdriver=vesa vga=788"
   fi
 
   list=`echo $list | tr ' ' '\n' | grep -vx $linux | tr '\n' ' '`
