diff -ur grub-2.00_orig/m4/gnulib-common.m4 grub-2.00/m4/gnulib-common.m4
--- grub-2.00_orig/m4/gnulib-common.m4	2010-12-01 14:45:43.000000000 +0000
+++ grub-2.00/m4/gnulib-common.m4	2013-02-23 23:42:24.977247685 +0000
@@ -129,7 +129,7 @@
     AC_SUBST([MKDIR_P])])], [
   dnl For autoconf < 2.60: Backport of AC_PROG_MKDIR_P.
   AC_DEFUN_ONCE([AC_PROG_MKDIR_P],
-    [AC_REQUIRE([AM_PROG_MKDIR_P])dnl defined by automake
+    [AC_REQUIRE([AC_PROG_MKDIR_P])dnl defined by automake
      MKDIR_P='$(mkdir_p)'
      AC_SUBST([MKDIR_P])])])
 
diff -ur grub-2.00_orig/m4/intl.m4 grub-2.00/m4/intl.m4
--- grub-2.00_orig/m4/intl.m4	2010-12-01 14:45:43.000000000 +0000
+++ grub-2.00/m4/intl.m4	2013-02-23 23:42:25.005247531 +0000
@@ -25,7 +25,7 @@
 AC_DEFUN([AM_INTL_SUBDIR],
 [
   AC_REQUIRE([AC_PROG_INSTALL])dnl
-  AC_REQUIRE([AM_PROG_MKDIR_P])dnl defined by automake
+  AC_REQUIRE([AC_PROG_MKDIR_P])dnl defined by automake
   AC_REQUIRE([AC_PROG_CC])dnl
   AC_REQUIRE([AC_CANONICAL_HOST])dnl
   AC_REQUIRE([gt_GLIBC2])dnl
diff -ur grub-2.00_orig/m4/po.m4 grub-2.00/m4/po.m4
--- grub-2.00_orig/m4/po.m4	2010-12-01 14:45:43.000000000 +0000
+++ grub-2.00/m4/po.m4	2013-02-23 23:42:25.012247493 +0000
@@ -24,7 +24,7 @@
 [
   AC_REQUIRE([AC_PROG_MAKE_SET])dnl
   AC_REQUIRE([AC_PROG_INSTALL])dnl
-  AC_REQUIRE([AM_PROG_MKDIR_P])dnl defined by automake
+  AC_REQUIRE([AC_PROG_MKDIR_P])dnl defined by automake
   AC_REQUIRE([AM_NLS])dnl
 
   dnl Release version of the gettext macros. This is used to ensure that
