diff --git a/po/ru.po b/po/ru.po
index 0be2287..a54b5ba 100644
--- a/po/ru.po
+++ b/po/ru.po
@@ -5282,12 +5282,12 @@ msgid "Found kernel module directory: %s\\n"
 msgstr "Найден каталог модулей ядра: %s\\n"
 
 #: util/grub.d/10_linux.in:89
-msgid "%s, with Linux %s (recovery mode)"
-msgstr "%s, с Linux %s (режим восстановления)"
+msgid "%s - Linux %s (recovery mode)"
+msgstr "%s - Linux %s (режим восстановления)"
 
 #: util/grub.d/10_linux.in:91
-msgid "%s, with Linux %s"
-msgstr "%s, с Linux %s"
+msgid "%s - Linux %s"
+msgstr "%s - Linux %s"
 
 #: util/grub.d/10_linux.in:135 util/grub.d/20_linux_xen.in:121
 msgid "Loading Linux %s ..."
diff --git a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
index a2060b3..c409af3 100644
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -86,9 +86,9 @@ linux_entry ()
   if [ x$type != xsimple ] ; then
       case $type in
 	  recovery)
-	      title="$(gettext_printf "%s, with Linux %s (recovery mode)" "${os}" "${version}")" ;;
+	      title="$(gettext_printf "%s - Linux %s (recovery mode)" "${os}" "${version}")" ;;
 	  *)
-	      title="$(gettext_printf "%s, with Linux %s" "${os}" "${version}")" ;;
+	      title="$(gettext_printf "%s - Linux %s" "${os}" "${version}")" ;;
       esac
       if [ x"$title" = x"$GRUB_ACTUAL_DEFAULT" ] || [ x"Previous Linux versions>$title" = x"$GRUB_ACTUAL_DEFAULT" ]; then
 	  replacement_title="$(echo "Advanced options for ${OS}" | sed 's,>,>>,g')>$(echo "$title" | sed 's,>,>>,g')"
diff --git a/util/grub.d/30_os-prober.in b/util/grub.d/30_os-prober.in
index 9294647..ceef906 100644
--- a/util/grub.d/30_os-prober.in
+++ b/util/grub.d/30_os-prober.in
@@ -149,7 +149,7 @@ for OS in ${OSPROBED} ; do
   case ${BOOT} in
     chain)
 
-	  onstr="$(gettext_printf "(on %s)" "${DEVICE}")"
+	  onstr="$(gettext_printf "(on %s)" "${DEVICE#/dev/}")"
       cat << EOF
 menuentry '$(echo "${LONGNAME} $onstr" | grub_quote)' --class windows --class os \$menuentry_id_option 'osprober-chain-$(grub_get_device_id "${DEVICE}")' {
 EOF
@@ -177,7 +177,9 @@ EOF
       boot_device_id=
       is_first_entry=true
       title_correction_code=
-      OS="${LONGNAME}"
+      onstr="$(gettext_printf "(on %s)" "${DEVICE#/dev/}")"
+      # List of items with menuentry names removed (to keep track of identical items)
+      linux_unique_ids="`echo \" ${LINUXPROBED}\" | sed 's/ \([^:]*:[^:]*:\)[^:]*:/ \1:/g'`"
 
       for LINUX in ${LINUXPROBED} ; do
         LROOT="`echo ${LINUX} | cut -d ':' -f 1`"
@@ -187,16 +189,11 @@ EOF
         LINITRD="`echo ${LINUX} | cut -d ':' -f 5`"
         LPARAMS="`echo ${LINUX} | cut -d ':' -f 6- | tr '^' ' '`"
 
-        if [ -z "${LLABEL}" ] ; then
-          LLABEL="${LONGNAME}"
-        fi
-
 	if [ "${LROOT}" != "${LBOOT}" ]; then
 	  LKERNEL="${LKERNEL#/boot}"
 	  LINITRD="${LINITRD#/boot}"
 	fi
 
-	onstr="$(gettext_printf "(on %s)" "${DEVICE}")"
 	recovery_params="$(echo "${LPARAMS}" | grep single)" || true
 	counter=1
 	while echo "$used_osprober_linux_ids" | grep 'osprober-gnulinux-$LKERNEL-${recovery_params}-$counter-$boot_device_id' > /dev/null; do
@@ -212,8 +209,13 @@ EOF
 	fi
 
 	if [ "x$is_first_entry" = xtrue ]; then
+            if [ -z "${LLABEL}" ] ; then
+                title="${LONGNAME} $onstr"
+            else
+                title="${LLABEL} $onstr"
+            fi
             cat << EOF
-menuentry '$(echo "$OS" | grub_quote)' --class ${CLASS} --class gnu-linux --class gnu --class os \$menuentry_id_option 'osprober-gnulinux-simple-$boot_device_id' {
+menuentry '$(echo "$title" | grub_quote)' --class ${CLASS} --class gnu-linux --class gnu --class os \$menuentry_id_option 'osprober-gnulinux-simple-$boot_device_id' {
 EOF
 	    save_default_entry | sed -e "s/^/\t/"
 	    add_options
@@ -229,10 +231,24 @@ EOF
         cat << EOF
 }
 EOF
-	    echo "submenu '$(gettext_printf "Advanced options for %s" "${OS}" | grub_quote)' \$menuentry_id_option 'osprober-gnulinux-advanced-$boot_device_id' {"
+	    echo "submenu '$(gettext_printf "Advanced options for %s" "$title" | grub_quote)' \$menuentry_id_option 'osprober-gnulinux-advanced-$boot_device_id' {"
 	    is_first_entry=false
+
+	    # If there are other entries in the list with absolutely the same kernel, parameters and initrd
+	    # we don't duplicate this main item in the Advanced submenu: it will appear there anyway only with
+	    # different title, and all chances are, this other title will be better suited (since it most probably
+	    # came from the Advanced submebu of the target system's Grub2 menu).
+	    unique_id="`echo ${LINUX} | sed 's/^\([^:]*:[^:]*:\)[^:]*:/\1:/'`"
+	    if ((`echo "${linux_unique_ids}" | grep -Fo "${unique_id}" | wc -l` > 1)); then
+	      continue
+	    fi
+	fi
+	VKERNEL=`echo $LKERNEL | sed -e 's,^\(/boot\)\?\(/vmlinuz-\|/vmlinux-\|/kernel-\),,'`
+	if [ -z "${LLABEL}" ] ; then
+		title="${LONGNAME} - Linux ${VKERNEL} $onstr"
+	else
+		title="${LLABEL} $onstr"
 	fi
-	title="${LLABEL} $onstr"
         cat << EOF
 	menuentry '$(echo "$title" | grub_quote)' --class ${CLASS} --class gnu-linux --class gnu --class os \$menuentry_id_option 'osprober-gnulinux-$LKERNEL-${recovery_params}-$boot_device_id' {
 EOF
@@ -251,7 +267,7 @@ EOF
 	}
 EOF
 	if [ x"$title" = x"$GRUB_ACTUAL_DEFAULT" ] || [ x"Previous Linux versions>$title" = x"$GRUB_ACTUAL_DEFAULT" ]; then
-	    replacement_title="$(echo "Advanced options for ${OS}" | sed 's,>,>>,g')>$(echo "$title" | sed 's,>,>>,g')"
+	    replacement_title="$(echo "Advanced options for ${LONGNAME}" | sed 's,>,>>,g')>$(echo "$title" | sed 's,>,>>,g')"
 	    quoted="$(echo "$GRUB_ACTUAL_DEFAULT" | grub_quote)"
 	    title_correction_code="${title_correction_code}if [ \"x\$default\" = '$quoted' ]; then default='$(echo "$replacement_title" | grub_quote)'; fi;"
 	    grub_warn "$(gettext_printf "Please don't use old title \`%s' for GRUB_DEFAULT, use \`%s' (for versions before 2.00) or \`%s' (for 2.00 or later)" "$GRUB_ACTUAL_DEFAULT" "$replacement_title" "gnulinux-advanced-$boot_device_id>gnulinux-$version-$type-$boot_device_id")"
@@ -268,7 +284,7 @@ EOF
       osx_entry xnu_kernel64 64
     ;;
     hurd)
-      onstr="$(gettext_printf "(on %s)" "${DEVICE}")"
+      onstr="$(gettext_printf "(on %s)" "${DEVICE#/dev/}")"
       cat << EOF
 menuentry '$(echo "${LONGNAME} $onstr" | grub_quote)' --class hurd --class gnu --class os \$menuentry_id_option 'osprober-gnuhurd-/boot/gnumach.gz-false-$(grub_get_device_id "${DEVICE}")' {
 EOF
