diff -Naur grub-2.00/include/grub/term.h grub-new/include/grub/term.h
--- grub-2.00/include/grub/term.h	2012-05-17 16:57:01.000000000 +0400
+++ grub-new/include/grub/term.h	2012-11-30 12:52:34.958060669 +0400
@@ -229,8 +229,8 @@
 };
 typedef struct grub_term_output *grub_term_output_t;
 
-#define GRUB_TERM_DEFAULT_NORMAL_COLOR 0x07
-#define GRUB_TERM_DEFAULT_HIGHLIGHT_COLOR 0x70
+#define GRUB_TERM_DEFAULT_NORMAL_COLOR 0x0F
+#define GRUB_TERM_DEFAULT_HIGHLIGHT_COLOR 0xF0
 #define GRUB_TERM_DEFAULT_STANDARD_COLOR 0x07
 
 extern struct grub_term_output *EXPORT_VAR(grub_term_outputs_disabled);
